src/Emogrifier/CssConcatenator.php

```diff
@@ -0,0 +1,156 @@
+<?php
+
+namespace Pelago\Emogrifier;
+
+/**
+ * Facilitates building a CSS string by appending rule blocks one at a time, checking whether the media query,
+ * selectors, or declarations block are the same as those from the preceding block and combining blocks in such cases.
+ *
+ * Example:
+ *  $concatenator = new CssConcatenator();
+ *  $concatenator->append(['body'], 'color: blue;');
+ *  $concatenator->append(['body'], 'font-size: 16px;');
+ *  $concatenator->append(['p'], 'margin: 1em 0;');
+ *  $concatenator->append(['ul', 'ol'], 'margin: 1em 0;');
+ *  $concatenator->append(['body'], 'font-size: 14px;', '@media screen and (max-width: 400px)');
+ *  $concatenator->append(['ul', 'ol'], 'margin: 0.75em 0;', '@media screen and (max-width: 400px)');
+ *  $css = $concatenator->closeBlocksAndGetCss();
+ *
+ * `$css` (if unminified) would contain the following CSS:
+ * ` body {
+ * `   color: blue;
+ * `   font-size: 16px;
+ * ` }
+ * ` p, ul, ol {
+ * `   margin: 1em 0;
+ * ` }
+ * ` @media screen and (max-width: 400px) {
+ * `   body {
+ * `     font-size: 14px;
+ * `   }
+ * `   ul, ol {
+ * `     margin: 0.75em 0;
+ * `   }
+ * ` }
```

Comment:
I think a plain string as data structure for this is not optimal. I propose putting this in an array (or object) structure that then is serialized to a string at the end. Like this:
[string media, string[] selectors, string[] rules]

This would also make the code more straightforward.

@JakeQZ What do you think?

Reply:
I don't see how this would help with the data structure given.  Wouldn't it just be in effect buffering the calls to `append()` to be replayed later, when the same processing would still be required?

If there was to be some kind of intermediate data structure, I think to be of benefit it would need to reflect the nested nature of the CSS, i.e. an array of media rules at the top level (including `""` for rules not within a media rule), and within each an array of rule blocks for selectors.  So in the example, after combining adjacent rules, we'd have something like
```
[
  [
    'media' => '',
    'rules' => [
      [
        'selectors' => ['body'],
        'declarations' => 'color: blue; font-size: 16px;',
      ],
      [
        'selectors' => ['p', 'ul', 'ol']
        'declarations' => 'margin: 1em 0;',
      ], 
    ],
  ],
  [
    'media' => '@media screen and (max-width: 400px)',
    'rules' => [
      [
        'selectors' => ['body'],
        'declarations' => 'font-size: 14px;'
      ],
      [
        'selectors' => ['ul', 'ol']
        'declarations' => 'margin: 0.75em 0;'
      ],
    ],
  ],
];
```

`append()` could then combine the rule being added with the last entry(ies) of the existing data where media/selectors/declarations match, and create a new entry where they don't.  I don't see any particular point deferring the combining of rules.

Does that sound okay?

Reply:
> I don't see how this would help with the data structure given. Wouldn't it just be in effect buffering the calls to append() to be replayed later, when the same processing would still be required?

The idea is that this will make the code straightforward: "adding CSS" means "sort it into the existing structure that is eminent in the code", and "getting the CSS" means "serializing the existing structure". So there is no need for the structure to have state (like "currently within a media query").

Reply:
> append() could then combine the rule being added with the last entry(ies) of the existing data where media/selectors/declarations match, and create a new entry where they don't. I don't see any particular point deferring the combining of rules.
>
> Does that sound okay?

Yes, that sounds perfect!